File: solutions/python/easy/p0344_reverse_string.py
"""LeetCode 344: Reverse String.

Problem: https://leetcode.com/problems/reverse-string/
"""

from typing import List


class Solution:
    def reverseString(self, s: List[str]) -> None:
        """Swap characters from both ends until the pointers meet.

        Time: O(n)
        Space: O(1)
        """
        left, right = 0, len(s) - 1
        while left < right:
            s[left], s[right] = s[right], s[left]
            left += 1
            right -= 1

File: solutions/java/easy/P0344_ReverseString.java
/**
 * LeetCode 344: Reverse String.
 *
 * Problem: https://leetcode.com/problems/reverse-string/
 *
 * Time complexity: O(n)
 * Space complexity: O(1)
 */
public class P0344_ReverseString {
    public void reverseString(char[] s) {
        int left = 0;
        int right = s.length - 1;

        while (left < right) {
            char temp = s[left];
            s[left] = s[right];
            s[right] = temp;

            left++;
            right--;
        }
    }

    public static void main(String[] args) {
        P0344_ReverseString solution = new P0344_ReverseString();

        char[] first = new char[] {'h', 'e', 'l', 'l', 'o'};
        solution.reverseString(first);
        printArray(first); // Expected: [o, l, l, e, h]

        char[] second = new char[] {'H', 'a', 'n', 'n', 'a', 'h'};
        solution.reverseString(second);
        printArray(second); // Expected: [h, a, n, n, a, H]
    }

    private static void printArray(char[] s) {
        System.out.print("[");
        for (int i = 0; i < s.length; i++) {
            if (i > 0) {
                System.out.print(", ");
            }
            System.out.print(s[i]);
        }
        System.out.println("]");
    }
}
